File: LeetCode/iter1/java/697.degree-of-an-array.java
package org.ironman.base.playground.leetcode;

import java.util.HashMap;
import java.util.Map;

public class Solution {

    public static void main(String[] args) {
        System.out.println(findShortestSubArray(new int[]{1, 2, 2, 3, 1}));
        System.out.println(findShortestSubArray(new int[]{1, 2, 2, 3, 1, 4, 2}));
        System.out.println(findShortestSubArray(new int[]{2, 3}));
        System.out.println(findShortestSubArray(new int[]{3}));
    }

    public static int findShortestSubArray(int[] nums) {
        Map<Integer, Integer> num2FirstIdx = new HashMap<>(), num2Cnt = new HashMap<>();
        int ret = Integer.MAX_VALUE, maxDegree = 0;

        for (int i = 0; i < nums.length; i++) {
            num2FirstIdx.putIfAbsent(nums[i], i);
            int cnt = num2Cnt.getOrDefault(nums[i], 0) + 1;
            num2Cnt.put(nums[i], cnt);

            if (maxDegree < cnt) {
                maxDegree = cnt;
                ret = i - num2FirstIdx.get(nums[i]) + 1;
            } else if (maxDegree == cnt) {
                ret = Math.min(ret, i - num2FirstIdx.get(nums[i]) + 1);
            }
        }

        return ret;
    }

}

File: LeetCode/iter1/c++/697.degree-of-an-array.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:
	int findShortestSubArray(vector<int>& nums) {
		unordered_map<int, int> num2FirstIdx, num2Cnt;
		int ret = INT_MAX, maxDegree = 0;
		for (int i = 0; i < nums.size(); i++) {
			if (num2FirstIdx.count(nums[i]) == 0) {
				num2FirstIdx[nums[i]] = i;
			}

			int degree = num2Cnt[nums[i]] + 1;
			num2Cnt[nums[i]] = degree;
			if (degree > maxDegree) {
				maxDegree = degree;
				ret = i - num2FirstIdx[nums[i]] + 1;
			}
			else if (degree == maxDegree) {
				ret = min(ret, i - num2FirstIdx[nums[i]] + 1);
			}
		}


		return ret;
	}
};

int main()
{
	Solution solution;
	vector<int> num1;
	num1.push_back(1);
	num1.push_back(2);
	num1.push_back(2);
	num1.push_back(3);
	num1.push_back(1);

	vector<int> num2;
	num2.push_back(1);
	num2.push_back(2);
	num2.push_back(2);
	num2.push_back(3);
	num2.push_back(1);
	num2.push_back(4);
	num2.push_back(2);

	vector<int> num3;
	num3.push_back(3);

	cout << solution.findShortestSubArray(num1) << endl;
	cout << solution.findShortestSubArray(num2) << endl;
	cout << solution.findShortestSubArray(num3) << endl;

	return 0;
}